### PR TITLE
WIP Victoria Metrics

### PR DESCRIPTION
**Pull request by kev306**

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -212,3 +213,19 @@ io.use(
 );
 
 socketServer(io);
+
+// Periodically push metrics every 15 seconds to VictoriaMetrics
+if (process.env.ENV === 'prod') {
```
Comment (vck3000) [inline src/app.ts:218]: Let's not restrict it to just prod. We wanna have this guy publishing for staging/local as well.

File: src/clients/victoriaMetrics/promAgent.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+export class PromAgent {
+  private metricNames: Set<string>;
+  private pushMetricsErrorsTimestamps: number[] = [];
+
+  constructor() {
+    this.metricNames = new Set<string>();
+  }
+
+  public addMetricName(metricName: string) {
+    assert(
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promAgent.ts:20]: Can you double check this completely takes down the server? And not just handled by the throw catch?

File: src/clients/victoriaMetrics/promAgent.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    this.metricNames.add(metricName);
+  }
+
+  // TODO-kev: Below is purely for testing. Keep or remove?
+  public getMetricNames(): Set<string> {
+    return this.metricNames;
+  }
+
+  public async pushMetricsToVictoriaMetrics() {
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promAgent.ts:33]:
```suggestion
  public async pushMetrics() {
```

File: src/clients/victoriaMetrics/promAgent.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      },
+    });
+
+    if (!response.ok) {
+      // TODO-kev: Remove below
+      console.log('ERROR PUSHING METRICS');
+
+      // Alert while errors are less than MAX_PUSH_METRICS_ERRORS
+      const now = Date.now();
+      this.pushMetricsErrorsTimestamps.push(now);
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promAgent.ts:50]: While the first entry is old, remove the first entry. Then check the size. Suppress if >5 or something. If alerting, push the timestamp up.

File: src/clients/victoriaMetrics/promAgent.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        );
+
+      // TODO-kev: Check below. Particularly error stack?. Consider what to do in cases where it exceeds
+      if (this.pushMetricsErrorsTimestamps.length <= MAX_PUSH_METRICS_ERRORS) {
+        sendToDiscordAdmins(
+          `Failed to push metrics: status=${response.status} text=${response.statusText}`,
+        );
+      }
+    } else {
+      this.pushMetricsErrorsTimestamps = [];
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promAgent.ts:63]: Let's rm the else.

File: src/clients/victoriaMetrics/promMetricGaugeFactory.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    } = {
+      name,
+      help,
+    };
+
+    if (collect) {
+      gaugeConfig.collect = collect;
+    }
+
+    this.gauge = new promClient.Gauge(gaugeConfig);
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promMetricGaugeFactory.ts:23]:
```suggestion
    this.gauge = new promClient.Gauge({ name, help, collect });
```

File: src/clients/victoriaMetrics/promMetricGaugeFactory.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    this.gauge.set(num);
+  }
+
+  public gaugeInc(num: number) {
+    this.gauge.inc(num);
+  }
+
+  public gaugeDec(num: number) {
+    this.gauge.dec(num);
+  }
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promMetricGaugeFactory.ts:36]: Let's rm for now. Add it when we need it.

File: src/sockets/sockets.ts [hunk 8]
```diff
@@ -60,6 +59,18 @@ if (process.env.NODE_ENV !== 'test') {
   }, 1000);
 }
 
+const onlinePlayersMetric = new PromMetricGaugeFactory(
```
Comment (vck3000) [inline src/sockets/sockets.ts:62]:
```suggestion
const onlinePlayersMetric = new PromMetricGauge(
```

File: src/clients/victoriaMetrics/promAgent.ts (hunk 2 above)
Comment (kev306) [reply src/clients/victoriaMetrics/promAgent.ts:17]: It does not take down server, changing to process.exit

File: src/clients/victoriaMetrics/promAgent.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+export class PromAgent {
+  private metricNames: Set<string>;
+  private pushMetricsErrorsTimestamps: number[] = [];
+
+  constructor() {
+    this.metricNames = new Set<string>();
+  }
+
+  public addMetricName(metricName: string) {
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promAgent.ts:15]:
```suggestion
  public registerMetric(metricName: string) {
```

File: src/clients/victoriaMetrics/promAgent.ts [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+import { sendToDiscordAdmins } from '../discord';
+
+const MAX_PUSH_METRICS_ERRORS = 5;
+const PUSH_METRICS_ERRORS_RATE_LIMIT = 60 * 60 * 1000; // 1 hour
+
+export class PromAgent {
+  private metricNames: Set<string>;
+  private pushMetricsErrorsTimestamps: number[] = [];
+
+  constructor() {
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promAgent.ts:11]: Let's add a function handler here in the constructor for this class to call when a duplicate metric is registered. 

Makes it easier for testing and makes the code cleaner.

File: src/clients/victoriaMetrics/promMetricGauge.ts [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+export class PromMetricGauge {
+  private gauge: Gauge;
+
+  constructor(gaugeConfig: GaugeConfig) {
+    promAgent.addMetricName(gaugeConfig.name);
+    this.gauge = new promClient.Gauge(gaugeConfig);
+  }
+
+  public set(val: number, labels?: Record<string, string>) {
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promMetricGauge.ts:19]: Let's add a comment above.

```
Be careful if you use labels. Should you really be using one?
```

File: .env.example [hunk 12]
```diff
@@ -11,8 +11,6 @@ SERVER_DOMAIN=127.0.0.1
```
Comment (vck3000) [inline .env.example:11]: Can u add to the readme as well to describe how to set up Grafana.

File: docker-compose.yml [hunk 13]
```diff
@@ -35,3 +35,28 @@ services:
...
+    image: grafana/grafana:latest
+    ports:
+      - 5000:3000
+    volumes:
+      - ./docker-compose-data/grafana:/var/lib/grafana
+    restart: always
+    environment:
+      GF_SECURITY_ADMIN_USER: ${GF_SECURITY_ADMIN_USER}
+      GF_SECURITY_ADMIN_PASSWORD: ${GF_SECURITY_ADMIN_PASSWORD}
+    user: "$UID:$GID" # TODO-kev: Warning where UID and GID are not set, defaults to blank string
```
Comment (vck3000) [inline docker-compose.yml:60]: Needed?

File: README.md [hunk 14]
```diff
@@ -55,6 +55,14 @@ $ yarn
 3. Create a Bucket named `proavalon` in MinIO.
 4. Set its Access Policy to Public.
 
+### Setting up Grafana
+
+1. Access Grafana via localhost:5000
+2. Connect `victoria-metrics` by following steps 3-5.
+3. From the main menu: Connections -> Data sources -> Add new data source.
```
Comment (vck3000) [inline README.md:62]:
```suggestion
    a. From the main menu: Connections -> Data sources -> Add new data source.
```

File: src/clients/victoriaMetrics/promAgent.ts [hunk 15]
```diff
@@ -33,7 +25,19 @@ export class PromAgent {
...
   public async pushMetrics() {
-    const metrics = await promClient.register.metrics(); // Will call any collect() functions for gauges
+    let metrics: string;
+
+    try {
+      metrics = await promClient.register.metrics(); // Will call any collect() functions for gauges
+    } catch (e) {
+      // Exit program if non-initialised labels are used in collect() function
+      if (e.message.includes('label')) {
+        console.error(e);
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promAgent.ts:35]: Send to discord admins

File: src/clients/victoriaMetrics/promMetricGauge.ts [hunk 16]
```diff
@@ -5,22 +5,15 @@ interface GaugeConfig {
   name: string;
   help: string;
   labelNames?: string[];
-  collect?: () => void;
+  collect?: () => void; // Refer to prom-client docs on how this should be configured
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promMetricGauge.ts:8]:
```suggestion
  collect?: () => void; // Refer to prom-client docs on how this should be used
```

File: src/clients/victoriaMetrics/tests/promAgent.test.ts [hunk 17]
```diff
@@ -4,13 +4,17 @@ describe('PromAgent', () => {
   let promAgent: PromAgent;
 
   beforeEach(() => {
-    promAgent = new PromAgent();
+    const dupeMetricErrorHandler = (metricName: string) => {
+      throw new Error(`Error metric name already exists: ${metricName}`);
+    };
```
Comment (vck3000) [inline src/clients/victoriaMetrics/tests/promAgent.test.ts:9]: nit: consider using a normal jest fn and expect it to be called.

File: src/clients/victoriaMetrics/tests/promAgent.test.ts [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+  });
+
+  it(`Adds unique metric names.`, () => {
+    expect(() => {
+      promAgent.registerMetric('metric_name_1');
+      promAgent.registerMetric('metric_name_2');
+    }).not.toThrow();
+
+    expect(promAgent.getMetricNames().has('metric_name_1'));
+    expect(promAgent.getMetricNames().has('metric_name_2'));
```
Comment (vck3000) [inline src/clients/victoriaMetrics/tests/promAgent.test.ts:21]: Remember let's not test implementation. We want to test behavior.

File: src/app.ts [hunk 19]
```diff
@@ -212,3 +213,19 @@ io.use(
 );
 
 socketServer(io);
+
+// Periodically push metrics every 15 seconds to VictoriaMetrics
+setInterval(async () => {
+  await promAgent.pushMetrics();
+}, 15000);
+
+// TODO-kev: Test - remove all lines below this once extracted out
```
Comment (vck3000) [inline src/app.ts:222]: Remove all the below here

File: src/clients/victoriaMetrics/promAgent.ts [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+        }
+      }
+
+      if (this.pushMetricsErrorsTimestamps.length <= MAX_PUSH_METRICS_ERRORS) {
+        sendToDiscordAdmins(errMsg);
+      }
+    }
+
+    // TODO-kev: Delete
+    console.log('Successfully pushed metrics.');
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promAgent.ts:69]: rm

File: src/clients/victoriaMetrics/promAgent.ts [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+      }
+
+      throw e;
+    }
+  }
+}
+
+const dupeMetricErrorHandler = (metricName: string) => {
+  console.error(`Error metric name already exists: ${metricName}`);
+  process.exit(1);
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promAgent.ts:93]: Discord error

File: src/sockets/sockets.ts [hunk 22]
```diff
@@ -60,6 +59,18 @@ if (process.env.NODE_ENV !== 'test') {
...
 
+const onlinePlayersMetric = new PromMetricGauge({
+  name: `online_players_total`,
+  help: `Number of online players`,
+  collect() {
+    // TODO-kev: uncomment below line once pushed
+    let randomInt = Math.floor(Math.random() * 40) + 1;
+    this.set(randomInt);
+
+    // this.set(allSockets.length);
```
Comment (vck3000) [inline src/sockets/sockets.ts:70]: fix this up

**Review by vck3000 — APPROVED** [review]

File: src/clients/victoriaMetrics/promAgent.ts [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+      // Exit program if non-initialised labels are used in collect() function
+      if (e.message.includes('label')) {
+        console.error(e);
+        process.exit(1);
+      }
+
+      throw e;
+    }
+
+    const response = await fetch(process.env.VM_IMPORT_PROMETHEUS_URL, {
```
Comment (vck3000) [inline src/clients/victoriaMetrics/promAgent.ts:37]: Can u test what happens if u don't include this in ur .env

`process.env.VM_IMPORT_PROMETHEUS_URL`

Comment (vck3000) [reply src/clients/victoriaMetrics/promAgent.ts:37]: Maybe just above it here, do a

```
if (process.env.VM_IMPORT_PROMETHEUS_URL === undefined) {
    process.exit(1);
}
```

**Pull request merged**